### ISSUE TITLE
how object segmentation are written in dataset json files?

### ISSUE THREAD

**Issue by dexter1608** [cocodataset/cocoapi#111]
hi,
I have polygons made using labelme annotation tool 
which are written like this 
points": [
        [
          258.69565217391306, 
          346.0869565217391
        ], 
        [
          252.60869565217394, 
          345.21739130434776
        ], 
        [
          245.6521739130435, 
          340.0
        ], 
        [
          243.04347826086956, 
          337.3913043478261
        ], 
        [
          239.56521739130437, 
          336.52173913043475
        ], 
        [
          232.60869565217394, 
          332.17391304347825
        ], 
        [
          231.73913043478262, 
          332.17391304347825
        ], 
        [
          222.17391304347825, 
          328.695652173913
        ], 
        [
          216.08695652173913, 
          320.86956521739125
        ], 
        [
          209.1304347826087, 
          319.13043478260863
        ], 
        [
          205.6521739130435, 
          318.2608695652174
        ], 
        [
          197.82608695652175, 
          315.6521739130435
        ], 
        [
          185.6521739130435, 
          315.6521739130435
        ], 
        [
          162.17391304347825, 
          308.695652173913
        ], 
        [
          142.17391304347825, 
          303.47826086956525
        ]]

and in coco 2014 dataset the segmentations are written like this 

{"segmentation": [[239.97,260.24,222.04,270.49,199.84,253.41,213.5,227.79,259.62,200.46,274.13,202.17,277.55,210.71,249.37,253.41,237.41,264.51,242.54,261.95,228.87,271.34]]

i dont know the difference how can I convert my polygons into coco type segmentation

thankyou in advance

**Comment by nightrome** [comment]
Just convert your own polygon representation to a binary mask (one per polygon) and then convert the mask to the COCO polygon format. E.g. the following script converts .png files (with non-overlapping polygons) https://github.com/nightrome/cocoapi/blob/master/PythonAPI/cocostuff/pngToCocoResultDemo.py

**Comment by priyanka-chaudhary** [comment]
@nightrome : I am using my own dataset and I have annotated the images. My groundtruth is an image of same size and for every pixel I have a number which is the class ID.

Like for Person class my ground truth image has pixel colour (1,1,1) same as COCO dataset. My question is if there are two person in an image should both be annotated with colour (1,1,1) or is there a different rule?
As to show two instances of an object we need some kind of distinction. Do you know how they are being annotated in the MS COCO dataset?

Any help would be really appreciated as I am not able to find this information anywhere. Thank you.

**Comment by nightrome** [comment]
That doesn't work. You cannot write instances (e.g. two persons) into an image and preserve that information. Take a look at the json format and then write your own script to create it: http://cocodataset.org/#download

**Comment by priyanka-chaudhary** [comment]
@nightrome 
I have checked the instances_val2017.json file. For example in image_id: 1000, there are multiple person. And in the file we have more than one person. Here is the part:

```{"segmentation": [[413.03,131.72,...,132.36]],"area": 1067.9535499999993,"iscrowd": 0,"image_id": 1000,"bbox": [405.93,120.42,37.13,45.52],"category_id": 1,"id": 1245349}```

```{"segmentation": [[277.59,392.37,...,380.01]],"area": 14372.617749999994,"iscrowd": 0,"image_id": 1000,"bbox": [265.33,95.86,88.92,315.88],"category_id": 1,"id": 1259139}```


```{"segmentation": [[281.11,418.74,...,423.72]],"area": 11511.253450000006,"iscrowd": 0,"image_id": 1000,"bbox": [209.23,174.64,99.63,249.08],"category_id": 1,"id": 1269164}```

Then my question is how did they get this from pixel-wise annotated image? Because if you assign different colour like (1,1,1) (2,2,2) (3,3,3) for different person in the image then in json format how do we reach to same category_id?

**Comment by nightrome** [comment]
The trick is to convert one object instance at a time from your format into a binary map and then into COCO polygon format. Also note that you don't need to make up an encoding into 3 rgb channels. You can just write images with a single channel.

**Comment by dexter1608** [comment]
@nightrome how to convert polygons into binary mask?

**Comment by nightrome** [comment]
https://github.com/nightrome/cocoapi/blob/master/PythonAPI/pycocotools/cocostuffhelper.py#L19

**Comment by priyanka-chaudhary** [comment]
@dexter1608 : I am trying to do that from masks. Not figured out yet.

**Comment by priyanka-chaudhary** [comment]
@nightrome : You mentioned ``` one object instance at a time from your format into a binary map and then into COCO polygon format```. I have extracted binary map but how to get COCO polygon format from that?

As when I passed binary mask to ```segmentationToCocoMask()``` I get this output: 

```[{'image_id': 9, 'category_id': 91, 'segmentation': {'counts': b'ga0:f8=dGC\\7\\1VMj20000M30000000000000000001O0000000000001O0000000O100001O00000000000000000000000000000000000000000000000001O000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000O11O00000NaKWK_4i4200000000001O1O0000000000000O1000000000000nKcKU3]4gLhKX3X4bLQL[3W5I5K3[MeIS2]6eMiI[2d60000000000000000000000000000000000O10000000000000000000000001O1O00000O100000000000000000000000000000000000000lM`M\\M`2d2dMXM\\2h2gMUMY2k2jMRMV2n2nMmLS2S3QNbLAjN^2d4bNcKS2]4nM^KW2a4Z100000000000000000000001O000000000000000000O1000000000000000000000000000000000000000000O0100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000O100000000000000000000000001N10000000000000000000000000000000000000000000000O1000O10000000000000000000000000000000000000000000000000000000000000000N2000000H;Gd0]L^JX2g6TN^>', 'size': [300, 416]}}] ```

**Comment by nightrome** [comment]
Looks good to me! From the [COCO page](http://cocodataset.org/#download):

> COCO provides segmentation masks for every object instance. This creates two challenges: storing masks compactly and performing mask computations efficiently. We solve both challenges using a custom Run Length Encoding (RLE) scheme. The size of the RLE representation is proportional to the number of boundaries pixels of a mask and operations such as area, union, or intersection can be computed efficiently directly on the RLE. Specifically, assuming fairly simple shapes, the RLE representation is O(√n) where n is number of pixels in the object, and common computations are likewise O(√n). Naively computing the same operations on the decoded masks (stored as an array) would be O(n).
> 
> The MASK API provides an interface for manipulating masks stored in RLE format. The API is defined below, for additional details see: MaskApi.m, mask.py, or MaskApi.lua. Finally, we note that a majority of ground truth masks are stored as polygons (which are quite compact), these polygons are converted to RLE when needed.

The RLE encoded polygons can be used (more or less) interchangeably with the more readable type of polygon annotation above.

**Comment by priyanka-chaudhary** [comment]
@nightrome : So can I use this above result to append to the ```instances_train2017.json``` and/or ```instances_val2017.json``` file(s) and use for training/validation?
Thanks a lot for all your help.

**Comment by nightrome** [comment]
Yes, we do that for the [COCO-Stuff annotations](http://images.cocodataset.org/annotations/stuff_annotations_trainval2017.zip).

**Comment by priyanka-chaudhary** [comment]
@nightrome : Yes I got the area, bbox from ```mask.py``` using encoded segmentation but I don't understand how ```id``` is generated. Is there a process for that too?

**Comment by nightrome** [comment]
AFAIK crowd sourced annotations have a big id and regular annotations a small id. But it should not matter as long as there are no duplicates.

**Comment by priyanka-chaudhary** [comment]
@nightrome : Any way to find out till where the ids are used or any documentation of it? as they are quite huge files not sure how to find that information.  I am not using crowd sourced annotations

**Comment by nightrome** [comment]
Generally speaking the [search function](https://github.com/cocodataset/cocoapi/search?utf8=%E2%9C%93&q=id&type=) is your friend. The codebase is not really big. I think it is only used as a unique index.

**Comment by dexter1608** [comment]
@nightrome 
the link you've given , I am failing to understand which function to use for polygon to binary mask?

**Comment by priyanka-chaudhary** [comment]
@nightrome : Thank you for the support. I found a way to do it.

**Comment by nightrome** [comment]
@dexter1608 
Oh I assumed that you would already have binary masks. I don't think such a function exists here. You could look for an external function like https://de.mathworks.com/help/images/ref/poly2mask.html
Otherwise try to figure out how COCO polygons work. From the comments in [mask.py](https://github.com/nightrome/cocoapi/blob/master/PythonAPI/pycocotools/mask.py):
```
#  poly    - Polygon stored as [[x1 y1 x2 y2...],[x1 y1 ...],...] (2D list)
# Both poly and bbs are 0-indexed (bbox=[0 0 1 1] encloses first pixel).
```

**Comment by raninbowlalala** [comment]
@priyanka-chaudhary How did you get the area, bbox from mask.py?

**Comment by raninbowlalala** [comment]
@priyanka-chaudhary Thank you very much. A further question, how did you generate the id? Which number should be selected to be the first number?

**Comment by priyanka-chaudhary** [comment]
@raninbowlalala  : I extracted all the ids from train and validation files and sorted them. The highest value I found for ```id``` is: **2232119**. You can take any number after that.

**Comment by raninbowlalala** [comment]
@priyanka-chaudhary Thanks for your help!

**Comment by raninbowlalala** [comment]
@nightrome I used  COCO-Stuff annotations to convert my own dataset to coco format, and added "area","bbox","id" to json file. When I finetuning the End-to-End_X-101-64x4d-FPN.pkl on my own dataset, I got the error like below:
E0227 16:22:45.168256 15153 pybind_state.h:422] Exception encountered running PythonOp function: ValueError: could not convert string to float: c

At:
  /home/liufang/detectron_venv/detectron/lib/utils/segms.py(134): polys_to_boxes
  /home/liufang/detectron_venv/detectron/lib/roi_data/mask_rcnn.py(46): add_mask_rcnn_blobs
  /home/liufang/detectron_venv/detectron/lib/roi_data/fast_rcnn.py(207): _sample_rois
  /home/liufang/detectron_venv/detectron/lib/roi_data/fast_rcnn.py(112): add_fast_rcnn_blobs
  /home/liufang/detectron_venv/detectron/lib/ops/collect_and_distribute_fpn_rpn_proposals.py(60): forward
terminate called after throwing an instance of 'caffe2::EnforceNotMet'
  what():  [enforce fail at pybind_state.h:423] . Exception encountered running PythonOp function: ValueError: could not convert string to float: c


It seems like the segmentation format is polygons, but my dataset used RLE format. Do you  have this problem?

**Comment by nightrome** [comment]
Sorry, but I am not familiar with that code. I suggest you contact the authors or look at how they parse the .json file. Or you save it as raw polygons instead of RLE format polygons.

**Comment by raninbowlalala** [comment]
@nightrome  OK, thanks for your help.

**Comment by raninbowlalala** [comment]
@nightrome Do you know how to convert mask binary image to raw polygons? Or convert RLE format to polygons format? Thanks a lot.

**Comment by nightrome** [comment]
Unfotunately not, I have never worked with polygons :-(

**Comment by raninbowlalala** [comment]
I found an issue and helped me convert mask binary image to polegon format. 
[https://github.com/facebookresearch/Detectron/issues/100](url)

**Comment by lscelory** [comment]
@priyanka-chaudhary I got the same output just like you after run `segmentationToCocoMask()`. I don't know what that means and how can I convert that string to polygon format. Do you have a solution? 
Look forward to your reply! Thanks a lot!

**Comment by nightrome** [comment]
@lscelory: The segmentationToCocoMask function is not part of this repository, but actually of the [COCO-Stuff repository](https://github.com/nightrome/cocoapi). Could you repost your question there with more information on what you are trying to achieve? I am losing track of the conversation here.

**Comment by priyanka-chaudhary** [comment]
@lscelory  : You can use it directly. You don't need to convert it to polygon format. Atleast I used it like that only. So I am not aware of any method to convert it to polygon format.

**Comment by nightrome** [comment]
@waspinator: Not sure why you would create an uncompressed RLE. For compressed RLE have a look at [the COCO-Stuff repository](https://github.com/nightrome/cocoapi/blob/master/PythonAPI/pycocotools/cocostuffhelper.py#L19). If you have any more questions please open a ticket there.

**Comment by nightrome** [comment]
@waspinator: I am not sure how to write them, but they can definitely be read using the standard mask API.  So imho there is no reason to store your annotations uncompressed. Just take whatever you have and compress it. Try to play around with [these functions](https://github.com/nightrome/cocoapi/blob/master/PythonAPI/pycocotools/mask.py#L37), that's all we have from Python.

**Comment by ambigus9** [comment]
@nightrome and @dexter1608 How to convert my own polygon representation to a binary mask?  or Segmentation format? Thanks!

**Comment by waspinator** [comment]
I wrote a library and article to help with creating COCO style datasets. 

https://patrickwasp.com/create-your-own-coco-style-dataset/

**Comment by waspinator** [comment]
@vsd550 did you try the reading the link above you?

**Comment by vsd550** [comment]
https://github.com/nightrome/cocoapi/blob/master/PythonAPI/pycocotools/cocostuffhelper.py#L19
In the above link as provided by @nightrome , what are the parameters to pass in the function **segmentationToCocoMask(labelMap, labelId)**.
I have created a binary mask of one object instance and now want the segmentation mask in the coco format.
@priyanka-chaudhary

**Comment by vsd550** [comment]
@waspinator  thanks for the answer. After I generate the binary mask 

I am using the function

> def polygons_to_mask(img_shape, polygons):
>   mask = np.zeros(img_shape[:2], dtype=np.uint8)
>   mask = Image.fromarray(mask)
>   xy = list(map(tuple, polygons))
>   ImageDraw.Draw(mask).polygon(xy=xy, outline=1, fill=1)
>   mask = np.array(mask, dtype=bool)
>   return mask

This function creates a binary mask given a polygon coordinates, one at a time.
Now how can I proceed to create the png and use your pycococreator tool
Thanks

**Comment by nirandiw** [comment]
What is the type of your input polygons?

**Comment by abhigoku10** [comment]
@nightrome @waspinator  hey guys , i am having a strange issue i.e i get mask mAP as 0 and bounding box mAP correct values . Here is the step i have performed 
1. i annotated a single image file using Labelme 
2.Converted the annotation to coco format 
3. Used this image as the GT 
4.Used a MaskRCNN on this image to obtain the output and calculate mAP by passing the annotations as GT
5. so Ideally since i am using the same image i should be getting mAP of box and mask as 100 
6. but i am getting box as 100 and mask as 0

**Comment by shubham-dpai** [comment]
> I wrote a library and article to help with creating COCO style datasets.
> 
> https://patrickwasp.com/create-your-own-coco-style-dataset/

Thanks a lot @waspinator !!! It helped me in annotation separation for crowded and non-crowded regions as well as clean decoding of masked data.

**Comment by LvJC** [comment]
A mask from png image's last(alpha) channel is a soft mask, which has good quality in edges and I think this is good for training. 
How to convert a soft mask into coco format label?
@nightrome @raninbowlalala @priyanka-chaudhary

**Comment by Dicko87** [comment]
Hi Guys, so I am wanting to do instance segmentation and the model that I am using expects the data to be in COCO 2017 format. When looking at the COCO 2017 json files for segmentation, I have noticed that  the segmentations in the json file are written as bounding box coordinates and not polygon coordinates. I was expecting polygon coordinates that would match up to points that create the bitmap images. I guess the mask images are enough information, coupled with the bounding box that encompasses the mask, for the model to understand what is going on ?

**Comment by sanghamitrajohri** [comment]
How to convert the masks to polygon format? @nightrome @ [priyanka-chaudhary](https://github.com/priyanka-chaudhary)
I have polygon annotated the file by labelme and got individual JSON files with respect to its image. And now I want to convert all the JSON files into the COCO Format.. like given here in the link   https://github.com/XinzeLee/PolygonObjectDetection/tree/main/coco20/annotations. So, is the repo valid for this one?

**Comment by Robotatron** [comment]
> I found an issue and helped me convert mask binary image to polygon format. Hope this answer will help others people. [https://github.com/facebookresearch/Detectron/issues/100](url)

@raninbowlalala 
![image](https://user-images.githubusercontent.com/95919873/210909106-689be993-b22d-4dc7-9a06-5ce781ee3612.png)
